### PR TITLE
*: append variadic warnings

### PR DESCRIPTION
**Pull request by cncal**
### What problem does this PR solve?

Problem Summary:
- Appending warnings to `StatementContext.mu.warnings` in a loop would cause frequent lock operations;
- [`AppendWarnings()`](https://github.com/pingcap/tidb/blob/cc3d9db37ca45437e2a44c297bfba03c22663f20/sessionctx/stmtctx/stmtctx.go#L404) without checking the length of `warns` would cause a wrong result when calling [`WarningCount`](https://github.com/pingcap/tidb/blob/cc3d9db37ca45437e2a44c297bfba03c22663f20/sessionctx/stmtctx/stmtctx.go#L363).

### What is changed and how it works?

What's Changed:
- Use variadic parameter `(warns ...error)` instead of `(warn error)`;
- Intercept `warns` if the length of warnings after appending is greater than `math.MaxUint16`.

### Check List 

Tests 

- Unit test
- Integration test

### Release note 
- no release note

### CODE REVIEW

**Review by time-and-fate — COMMENTED** [review]
Thanks for your contributions.

I think changes in `go.mod` and `go.sum` are irrelevant.
And for the second problem you mentioned, we need a test case for that. You can add a unit test.

**Comment by cncal** [comment]
@time-and-fate It's just a simple bugfix and I don't think test case is necessary.

File: sessionctx/stmtctx/stmtctx.go [hunk 1]
```diff
@@ -391,22 +391,27 @@ func (sc *StatementContext) SetWarnings(warns []SQLWarn) {
...
+	sc.AppendWarnings(ws)
 }
 
 // AppendWarnings appends some warnings.
 func (sc *StatementContext) AppendWarnings(warns []SQLWarn) {
 	sc.mu.Lock()
-	if len(sc.mu.warnings) < math.MaxUint16 {
-		sc.mu.warnings = append(sc.mu.warnings, warns...)
+	defer sc.mu.Unlock()
+	remainingLen := math.MaxUint16 - len(sc.mu.warnings)
```
Comment (zz-jason) [inline sessionctx/stmtctx/stmtctx.go:410]: what does magic number `math.MaxUint16` stand for, the max number of warnings for a query?

File: sessionctx/stmtctx/stmtctx.go [hunk 2]
```diff
@@ -391,22 +391,27 @@ func (sc *StatementContext) SetWarnings(warns []SQLWarn) {
...
-	sc.mu.Unlock()
+	ws := make([]SQLWarn, 0, len(warns))
+	for _, w := range warns {
+		ws = append(ws, SQLWarn{WarnLevelWarning, w})
+	}
+	sc.AppendWarnings(ws)
 }
 
 // AppendWarnings appends some warnings.
 func (sc *StatementContext) AppendWarnings(warns []SQLWarn) {
```
Comment (zz-jason) [inline sessionctx/stmtctx/stmtctx.go:407]: please add a UT for this bug-fix, no matter how small it is.

**Pull request closed without merging**